### PR TITLE
Feature/1.2.5.2b/service caetgory redo

### PR DESCRIPTION
**Pull request by jnakaso**
* created a hook for constants
* updated forms to use the app constants

### CODE REVIEW

File: src/services/useAppConstants.tsx [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    const [data, setData] = useState<AppConstant[]>([]);
+
+    useEffect(() => {
+        const fetchData = async () => {
+            setStatus('fetching');
+            if (!init) {
+                const response = await supabaseClient.from('app_constants')
+                    .select()
+                const types = new Set(response.data.map(c => c.type));
+                types.forEach(t => cache[t] = response.data.filter(c => c.type === t));
```
Comment (sharrief) [inline src/services/useAppConstants.tsx:23]: I think lines 22 and 23 could be written as one loop over the data to be a little more clear of the intention 

```
const cache: new Map<string, AppConstant[]>()
...
response.data.forEach((c) => {
    if (!cache.has(c.type)) cach.set(c.type, [])
    cache.get(c.type).push(c)
}
...
const data = cache.get(type)
```

File: __tests__/src/services/useAppConstants.test.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+            .mockReturnValue(Promise.resolve(response))
+
+
+        const { data: actual, status } = useAppConstants('TYPE')
+        expect(fromSpy).toHaveBeenCalledWith('app_constants')
+        expect(selectSpy).toHaveBeenCalled()
+        expect(actual).toEqual(expected)
+        expect(status).toEqual('STATUS')
+    });
+
```
Comment (sharrief) [inline __tests__/src/services/useAppConstants.test.ts:44]: I felt a litte fricion in trying to follow the intention of the test code, can we add few comments? 
For example the 3 calls useState are order dependent (this is a little fragile) and it was hard to parse where each var was used in a mock. 

Might help also to place the vars by the mock that uses them ("response" can be an inline var for the resovled value of .select). 

And since mockQueryBuilder.select is already a mock, you can call mockResovledValue directly on it.

File: __tests__/src/services/useAppConstants.test.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
+import React from 'react';
+import { useAppConstants } from '../../../src/services/useAppConstants';
+import supabaseClient from '../../../utils/supabaseClient';
+
+describe('useAppConstants tests', () => {
+
+    const mockFilterBuilder = {
+        eq: jest.fn(() => Promise.resolve({}))
+    };
```
Comment (sharrief) [inline __tests__/src/services/useAppConstants.test.ts:9]: I think this function can be deleted.

Comment (jnakaso) [reply __tests__/src/services/useAppConstants.test.ts]: ok.  Let's consider creating an Util test class.

File: __tests__/src/services/useAppConstants.test.ts (hunk 2 above)
Comment (jnakaso) [reply __tests__/src/services/useAppConstants.test.ts:28]: got it.  I'm redoing it how this is done.  I don't like it either.

File: src/services/useAppConstants.tsx (hunk 1 above)
Comment (jnakaso) [reply src/services/useAppConstants.tsx]: ok

File: src/services/AppService.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+  getAppConstants = async (type: string): Promise<AppConstant[]> => {
+    if (this.cache === null) {
+      return this.lookup()
+        .then(m => {
+          this.cache = m;
+          return this.cache.get(type);
+        })
+    } else {
+      return Promise.resolve(this.cache.get(type));
```
Comment (sharrief) [inline src/services/AppService.ts:39]: Don't need to wrap the value in a Promise as the function is marked async which wraps the value automatically

Comment (jnakaso) [reply src/services/AppService.ts:39]: thx

**Review by sharrief — APPROVED** [review]
Thanks!
